typescript/cli/src/rebalancer/monitor/Monitor.ts

```diff
@@ -0,0 +1,148 @@
+  }
+
+  async start() {
+    if (this.isMonitorRunning) {
+      // Cannot start the same monitor multiple times
+      this.emitter.emit(
+        MonitorEventType.Error,
+        new MonitorStartError('Monitor already running'),
+      );
+      return;
+    }
+
+    try {
+      this.isMonitorRunning = true;
+      monitorLogger.debug(
+        { checkFrequency: this.checkFrequency },
+        'Monitor started',
+      );
+      this.emitter.emit(MonitorEventType.Start);
+
+      while (this.isMonitorRunning) {
+        try {
+          monitorLogger.debug('Polling cycle started');
+          const event: MonitorEvent = {
+            tokensInfo: [],
+          };
+
+          for (const token of this.warpCore.tokens) {
+            monitorLogger.debug(
+              {
+                chain: token.chainName,
+                tokenSymbol: token.symbol,
+                tokenAddress: token.addressOrDenom,
+              },
+              'Checking token',
+            );
+            const bridgedSupply = await this.getTokenBridgedSupply(token);
+
+            event.tokensInfo.push({
+              token,
+              bridgedSupply,
+            });
+          }
+
+          // Emit the event warp routes info
+          this.emitter.emit(MonitorEventType.TokenInfo, event);
+          monitorLogger.debug('Polling cycle completed');
+        } catch (error) {
+          this.emitter.emit(
+            MonitorEventType.Error,
+            new MonitorPollingError(
+              `Error during monitor execution cycle: ${(error as Error).message}`,
+              error as Error,
+            ),
+          );
+        }
+
+        // Wait for the specified check frequency before the next iteration
+        await sleep(this.checkFrequency);
+      }
+    } catch (error) {
+      this.emitter.emit(
+        MonitorEventType.Error,
+        new MonitorStartError(
+          `Error starting monitor: ${(error as Error).message}`,
+          error as Error,
+        ),
+      );
+    }
+  }
+
+  private async getTokenBridgedSupply(
+    token: Token,
+  ): Promise<bigint | undefined> {
+    if (!token.isHypToken()) {
+      monitorLogger.warn(
+        {
+          chain: token.chainName,
+          tokenSymbol: token.symbol,
+          tokenAddress: token.addressOrDenom,
+        },
+        'Cannot get bridged balance for a non-Hyperlane token',
+      );
+      return;
+    }
+
+    const adapter = token.getHypAdapter(this.warpCore.multiProvider);
+    const bridgedSupply = await adapter.getBridgedSupply();
+
+    if (bridgedSupply === undefined) {
+      monitorLogger.warn(
+        {
+          chain: token.chainName,
+          tokenSymbol: token.symbol,
+          tokenAddress: token.addressOrDenom,
+        },
+        'Bridged supply not found for token',
+      );
+    }
+
+    return bridgedSupply;
+  }
+
+  stop() {
+    this.isMonitorRunning = false;
+    monitorLogger.info('Monitor stopped');
+    this.emitter.removeAllListeners();
+  }
```

Comment:
Shouldn't we make this dependent on if there is code executing in the while loop in the start method? like wait until the current loop execution ends and then gracefully exit instead of just shutting the monitor down? Although maybe as this is only doing work on metrics it might be not that important

Reply:
Strictly speaking, we're not "interrupting" the while loop as the current iteration will continue to the end. The more subtle effect here is that we're removing the listeners immediately, so the while loop won't be sending any new events after the call to `stop()`.

You may argue this behavior is desired as we may be doing other clean-up and may no longer want to process the events. However, our code emits all errors as events,  including any errors that happen during termination. So maybe we should think twice and decide not to remove the listeners here.